tuta-sdk/rust/src/crypto_entity_client.rs

```diff
@@ -0,0 +1,210 @@
+use std::sync::Arc;
+use serde::Deserialize;
+use crate::entity_client::IdType;
+#[mockall_double::double]
+use crate::entity_client::EntityClient;
+use crate::ApiCallError;
+#[mockall_double::double]
+use crate::crypto::crypto_facade::CryptoFacade;
+use crate::entities::Entity;
+use crate::entities::entity_facade::EntityFacade;
+#[mockall_double::double]
+use crate::instance_mapper::InstanceMapper;
+
+
+// A high level interface to manipulate encrypted entities/instances via the REST API
+pub struct CryptoEntityClient {
+    entity_client: Arc<EntityClient>,
+    entity_facade: Arc<EntityFacade>,
+    crypto_facade: Arc<CryptoFacade>,
+    instance_mapper: Arc<InstanceMapper>,
+}
+
+impl CryptoEntityClient {
+    pub fn new(
+        entity_client: Arc<EntityClient>,
+        entity_facade: Arc<EntityFacade>,
+        crypto_facade: Arc<CryptoFacade>,
+        instance_mapper: Arc<InstanceMapper>,
+    ) -> Self {
+        CryptoEntityClient { entity_client, entity_facade, crypto_facade, instance_mapper }
+    }
+    pub async fn load<T: Entity + Deserialize<'static>>(
+        &self,
+        id: &IdType,
+    ) -> Result<T, ApiCallError> {
+        let type_ref = T::type_ref();
+        let type_model = self.entity_client.get_type_model(&type_ref)?;
+        let mut parsed_entity = self.entity_client.load(&type_ref, id).await?;
+
+        if type_model.encrypted {
+            let possible_session_key = self.crypto_facade
+                .resolve_session_key(&mut parsed_entity, type_model)
+                .map_err(
+                    |error| ApiCallError::InternalSdkError {
```

Comment:
I am not sure if we need a more specific error about key resolution at some but it we can probably keep it like it is for now

Reply:
That is true, we should have more detail in the error. I have pushed a new commit now.